webui/react/src/stores/projects.tsx

```diff
@@ -5,121 +5,76 @@ import { getWorkspaceProjects } from 'services/api';
 import { Project } from 'types';
 import handleError from 'utils/error';
 import { Loadable, Loaded, NotLoaded } from 'utils/loadable';
+import { observable, WritableObservable } from 'utils/observable';
 
-type UpdateProjectsByIndex = (fn: (ws: Map<string, Project[]>) => Map<string, Project[]>) => void;
+class ProjectService {
+  projects: WritableObservable<Map<number, Project>> = observable(Map());
+  projectsByIndex: WritableObservable<Map<string, Project[]>> = observable(Map());
 
-type ProjectsContext = {
-  projects: Map<number, Project>;
-  projectsByIndex: Map<string, Project[]>;
-  updateProjects: (fn: (project: Map<number, Project>) => Map<number, Project>) => void;
-  updateProjectsByIndex: UpdateProjectsByIndex;
-};
-
-const ProjectsContext = createContext<ProjectsContext | null>(null);
+  canceler: AbortController;
 
-export const ProjectsProvider: React.FC<PropsWithChildren> = ({ children }) => {
-  const [workspaceProjects, setWorkspaceProjects] = useState<Map<string, Project[]>>(Map());
-  const [projects, setProjects] = useState<Map<number, Project>>(() => Map());
-
-  return (
-    <ProjectsContext.Provider
-      value={{
-        projects,
-        projectsByIndex: workspaceProjects,
-        updateProjects: setProjects,
-        updateProjectsByIndex: setWorkspaceProjects,
-      }}>
-      {children}
-    </ProjectsContext.Provider>
-  );
-};
+  constructor(canceler: AbortController) {
+    this.canceler = canceler;
+  }
 
-export const useFetchWorkspaceProjects = (
-  canceler: AbortController,
-): ((workspaceId: number) => Promise<void>) => {
-  const context = useContext(ProjectsContext);
+  fetchWorkspaceProjects = async (workspaceId: number): Promise<void> => {
+    try {
+      const response = await getWorkspaceProjects(
+        {
+          id: workspaceId,
+          limit: 0,
+        },
+        { signal: this.canceler.signal },
+      );
+      this.projectsByIndex.update((prevState: Map<string, Project[]>) => {
+        return prevState.set(`byworkspace-${workspaceId}`, response.projects);
+      });
+      this.projects.update((prevState: Map<number, Project>) => {
+        return prevState.withMutations((state: Map<number, Project>) => {
+          response.projects.forEach((proj) => state.set(proj.id, proj));
+        });
+      });
+    } catch (e) {
+      handleError(e);
+    }
+  };
+}
 
-  if (context === null) {
-    throw new Error('Attempted to use useFetchWorkspaceProjects outside of Projects Context');
-  }
+const ProjectsContext = createContext<ProjectService | null>(null);
 
-  const { updateProjectsByIndex, updateProjects } = context;
+export const ProjectsProvider: React.FC<PropsWithChildren> = ({ children }) => {
+  const [projectStore] = useState(() => new ProjectService(new AbortController()));
```

Comment:
non-blocking: it looks like there it not anywhere we are calling `canceler.abort()` on this controller. maybe a `useEffect(..., [])` to cancel when the provider unmounts?